File: engine/core/state.py
"""Runtime state model for deterministic tactical simulations."""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Dict, List, Optional, Tuple


@dataclass
class UnitState:
    unit_id: str
    team: str
    hp: int
    max_hp: int
    x: int
    y: int
    initiative: int
    attack_mod: int
    ac: int
    damage: str
    temp_hp: int = 0
    temp_hp_source: Optional[str] = None
    temp_hp_owner_effect_id: Optional[str] = None
    attack_damage_type: str = "physical"
    fortitude: int = 0
    reflex: int = 0
    will: int = 0
    actions_remaining: int = 3
    reaction_available: bool = True
    conditions: Dict[str, int] = field(default_factory=dict)
    condition_immunities: List[str] = field(default_factory=list)
    resistances: Dict[str, int] = field(default_factory=dict)
    weaknesses: Dict[str, int] = field(default_factory=dict)
    immunities: List[str] = field(default_factory=list)

    @property
    def alive(self) -> bool:
        return self.hp > 0


@dataclass
class MapState:
    width: int
    height: int
    blocked: List[Tuple[int, int]] = field(default_factory=list)


@dataclass
class EffectState:
    effect_id: str
    kind: str
    source_unit_id: Optional[str]
    target_unit_id: Optional[str]
    payload: Dict[str, object] = field(default_factory=dict)
    duration_rounds: Optional[int] = None
    tick_timing: Optional[str] = None


@dataclass
class BattleState:
    battle_id: str
    seed: int
    round_number: int
    turn_index: int
    turn_order: List[str]
    units: Dict[str, UnitState]
    battle_map: MapState
    effects: Dict[str, EffectState] = field(default_factory=dict)
    flags: Dict[str, bool] = field(default_factory=dict)
    event_sequence: int = 0

    @property
    def active_unit_id(self) -> str:
        return self.turn_order[self.turn_index]

    @property
    def active_unit(self) -> UnitState:
        return self.units[self.active_unit_id]

File: old/src/engine/state.ts
/**
 * Runtime state model for deterministic tactical simulations.
 * Mirrors engine/core/state.py
 */

export interface UnitState {
  unitId: string;
  team: string;
  hp: number;
  maxHp: number;
  x: number;
  y: number;
  initiative: number;
  attackMod: number;
  ac: number;
  damage: string;
  tempHp: number;
  tempHpSource: string | null;
  tempHpOwnerEffectId: string | null;
  attackDamageType: string;
  attackDamageBypass: string[];
  fortitude: number;
  reflex: number;
  will: number;
  actionsRemaining: number;
  reactionAvailable: boolean;
  speed: number;
  /** Melee reach in tiles (PF2e default = 1; reach weapons = 2). */
  reach: number;
  /** Attacks made this turn — drives Multiple Attack Penalty. Reset to 0 on turn start. */
  attacksThisTurn: number;
  conditions: Record<string, number>;
  conditionImmunities: string[];
  resistances: Record<string, number>;
  weaknesses: Record<string, number>;
  immunities: string[];
  /** Content-entry ids this unit may use. Undefined = all entries available. */
  abilities?: string[];
  /** Remaining uses for limited-use content entries (keyed by entry id). Absent = unlimited. */
  abilitiesRemaining: Record<string, number>;
}

export function unitAlive(unit: UnitState): boolean {
  return unit.hp > 0;
}

export interface MapState {
  width: number;
  height: number;
  blocked: Array<[number, number]>;
  /** Per-tile movement cost keyed by "x,y"; absent = 1 (normal). Difficult terrain = 2. */
  moveCost?: Record<string, number>;
  /** Per-tile cover grade keyed by "x,y"; absent = 0. 1 = standard, 2 = greater. */
  coverGrade?: Record<string, number>;
  /** Per-tile elevation keyed by "x,y"; absent = 0 (ground level). */
  elevation?: Record<string, number>;
}

export interface EffectState {
  effectId: string;
  kind: string;
  sourceUnitId: string | null;
  targetUnitId: string | null;
  payload: Record<string, unknown>;
  durationRounds: number | null;
  tickTiming: "turn_start" | "turn_end" | null;
}

export interface BattleState {
  battleId: string;
  seed: number;
  roundNumber: number;
  turnIndex: number;
  turnOrder: string[];
  units: Record<string, UnitState>;
  battleMap: MapState;
  effects: Record<string, EffectState>;
  flags: Record<string, boolean>;
  eventSequence: number;
}

export function activeUnitId(state: BattleState): string {
  return state.turnOrder[state.turnIndex];
}

export function activeUnit(state: BattleState): UnitState {
  return state.units[activeUnitId(state)];
}
